go-controller/pkg/ovn/controller/egressfirewall/egressfirewall.go

```diff
@@ -188,120 +290,120 @@ func (oc *DefaultNetworkController) syncEgressFirewall(egressFirewalls []interfa
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		// delete case
+		klog.V(3).Infof("Removing egress firewall %s", key)
+		p := libovsdbops.GetPredicate[*nbdb.ACL](oc.GetEgressFirewallACLDbIDsNoRule(namespace), nil)
+		invalidACLs, err := libovsdbops.FindACLsWithPredicate(oc.nbClient, p)
+		if err != nil {
+			return fmt.Errorf("error finding ACLs for egress firewall %s: %w", key, err)
+		}
+		pgPredicate := func(item *nbdb.PortGroup) bool {
+			if len(item.ACLs) == 0 {
+				return false
 			}
-			swWithACLsPred := func(sw *nbdb.LogicalSwitch) bool {
-				return aclsToDelete.HasAny(sw.ACLs...)
+			if item.ExternalIDs[libovsdbops.ObjectNameKey.String()] == namespace &&
+				item.ExternalIDs[libovsdbops.OwnerTypeKey.String()] == libovsdbops.NamespaceOwnerType {
+				return true
 			}
-			return libovsdbops.RemoveACLsFromLogicalSwitchesWithPredicate(oc.nbClient, swWithACLsPred, batchACLs...)
-		})
+			return false
+		}
+		pgs, err := libovsdbops.FindPortGroupsWithPredicate(oc.nbClient, pgPredicate)
 		if err != nil {
-			return fmt.Errorf("failed to remove egress firewall acls from node logical switches: %v", err)
+			return fmt.Errorf("error finding port groups for egress firewall namespace %s: %w", namespace, err)
+		}
+		pgNames := make([]string, 0, len(pgs))
+		for _, pg := range pgs {
+			pgNames = append(pgNames, pg.Name)
+		}
+		if err := libovsdbops.DeleteACLsFromPortGroups(oc.nbClient, pgNames, invalidACLs...); err != nil {
+			return fmt.Errorf("error deleting stale ACLs for egress firewall %s: %w", key, err)
+		}
+		if err := oc.dnsNameResolver.Delete(namespace); err != nil {
+			return err
+		}
+		numRules, ok := oc.ruleCounter.Load(key)
+		if ok {
+			metrics.UpdateEgressFirewallRuleCount(float64(-numRules.(uint32)))
+			oc.ruleCounter.Delete(key)
+		} else {
+			klog.Errorf("Unable to decrement egress firewall rule count, cache miss for key: %s", key)
 		}
+		metrics.DecrementEgressFirewallCount()
+		return nil
 	}
-	return nil
-}
 
-// moveACLsToNamespacedPortGroups syncs db from the previous version where all ACLs were attached to the ClusterPortGroup
-// to the new version where ACLs are attached to the namespace port groups.
-func (oc *DefaultNetworkController) moveACLsToNamespacedPortGroups(existingEFNamespaces map[string]bool, efACLs []*nbdb.ACL) error {
-	// find stale ACLs attached to a cluster port group, and move them to namespaced port groups
-	clusterPG, err := libovsdbops.GetPortGroup(oc.nbClient, &nbdb.PortGroup{
-		Name: oc.getClusterPortGroupName(types.ClusterPortGroupNameBase),
-	})
+	// add/update case
+	defer func() {
+		if statusErr := oc.setEgressFirewallStatus(ef, updateErr); statusErr != nil {
+			updateErr = utilerrors.Join(updateErr, fmt.Errorf("failed to update egress firewall status %s, error: %w",
+				GetEgressFirewallNamespacedName(ef), statusErr))
+		}
+	}()
+
+	pgName, err := oc.getNamespacePortGroupName(ef.Namespace)
```

Comment:
do you want to handle `util.IsInvalidPrimaryNetworkError(err)` separately to stop retrying, since it won't be fixed until NAD update happens?

Reply:
I worry a bit about that since the callback does not retry upon error, even though the error probability of being able to list in the informer cache is pretty much 0. The rate limiter should throttle back retries on the object so I don't think it is worth changing this.